### PR TITLE
Fix Drag & Drop JS runtime

### PR DESCRIPTION
**Pull request by jakubpeleska**
# Description

This PR completes the drag-and-drop support, which was previously implemented in:

- https://github.com/wailsapp/wails/pull/3203
- https://github.com/wailsapp/wails/pull/3250

## Changes
- drag and drop refactor - `draganddrop.js`
- fix for CSS class `wails-drop-target-active`
- add debounce on `dragleave` to avoid flickering of the `wails-drop-target-active` class

This implementation provides developers with enough functionality to start using this feature in production. Therefore, we believe this commit can conclude all previous efforts and mark the drag-and-drop feature as ready for production.

## How Has This Been Tested?
Fix was tested on a dummy project available on https://github.com/beam-transfer/wails-drag-and-drop
- [ ] Windows
- [x] macOS
- [x] Linux

### Video

- **red area** - drop zone `"--wails-drop-target": "drop"`
- **yellow area** - drop zone disabled `"--wails-drop-target": "none"`
- **teal area** - element without any special treatment

https://github.com/wailsapp/wails/assets/48656252/e87cbb2b-013a-408d-b6cf-cf8b26d4e0b0

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# Checklist:

- [x] I have updated `website/src/pages/changelog.mdx` with details of this PR
- [x] My code follows the general coding style of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by pavelbinar** [comment]
Hi @lyimmi , @APshenkin please take a look.

File: website/docs/reference/options.mdx [hunk 1]
```diff
@@ -64,6 +64,12 @@ func main() {
           UniqueId:               "c9c8fd93-6758-4144-87d1-34bdb0a8bd60",
           OnSecondInstanceLaunch: app.onSecondInstanceLaunch,
         },
+        DragAndDrop: &options.DragAndDrop{
+          EnableFileDrop:  false,
+          Disable:         false,
```
Comment (leaanthony) [inline website/docs/reference/options.mdx:69]:
```suggestion
          DisableWebViewDrop:         false,
```

**Review by leaanthony — COMMENTED** [review]
Thanks for taking the time to loop back to this 🙏 
What do you think the challenges for this approach might be for a multi-window application?

**Comment by APshenkin** [comment]
@jakubpeleska @pavelbinar 
This looks great. Thank you for finishing this!

Hope to see this released in v2 soon!

**Comment by jakubpeleska** [comment]
@leaanthony Thanks for taking a look.

> Thanks for taking the time to loop back to this 🙏 What do you think the challenges for this approach might be for a multi-window application?

As for the multi-window challenges, I can't really speak for the native implementations. However, the runtime JS should work out of the box, assuming that there are separate JS runtimes for each window. That being said, I did not test it, so I can't be sure. @APshenkin and @lyimmi might have more insight into the challenges of native implementations.

**Review by 9dl — APPROVED** [review]

File: v2/internal/frontend/desktop/darwin/WailsWebView.m [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    return [super performDragOperation: sender];
+
+  // getting all NSURL types
+  NSArray<Class> *url_class = @[[NSURL class]];
+  NSDictionary *options = @{};
+  NSArray<NSURL*> *files = [pboard readObjectsForClasses:url_class options:options];
+
+  // collecting all file paths
+  NSString *files_str = @"";
+  NSMutableArray *files_strs = [[NSMutableArray alloc] init];
```
Comment (leaanthony) [inline v2/internal/frontend/desktop/darwin/WailsWebView.m:42]: Does this need to be released?

Comment (jakubpeleska) [reply v2/internal/frontend/desktop/darwin/WailsWebView.m:42]: @APshenkin

Comment (APshenkin) [reply v2/internal/frontend/desktop/darwin/WailsWebView.m:42]: @leaanthony @jakubpeleska 

Here we iterate over NSURL array to convert them to just strings, then join then all in one string. 

So `files_strs` is needed. Only if someone have better know how to do this mapping, as I'm not Objective-C expert 😅. 

However `files_str` is redundant, as it's not used. You can remove it.

Comment (leaanthony) [reply v2/internal/frontend/desktop/darwin/WailsWebView.m:42]: I was thinking the `[alloc]` would require a `[release]`

File: v2/internal/frontend/desktop/darwin/WailsWebView.m [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  if( !types )
+    return [super performDragOperation: sender];
+
+  // getting all NSURL types
+  NSArray<Class> *url_class = @[[NSURL class]];
+  NSDictionary *options = @{};
+  NSArray<NSURL*> *files = [pboard readObjectsForClasses:url_class options:options];
+
+  // collecting all file paths
+  NSString *files_str = @"";
```
Comment (leaanthony) [inline v2/internal/frontend/desktop/darwin/WailsWebView.m:41]:
```suggestion
```

File: v2/internal/frontend/desktop/darwin/WailsWebView.m [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+  NSMutableArray *files_strs = [[NSMutableArray alloc] init];
+  for (NSURL *url in files)
+  {
+    const char *fs_path = [url fileSystemRepresentation];  //Will be UTF-8 encoded
+    NSString *fs_path_str = [[NSString alloc] initWithCString:fs_path encoding:NSUTF8StringEncoding];
+    [files_strs addObject:fs_path_str];
+//     NSLog( @"performDragOperation: file path: %s", fs_path );
+  }
+
+  NSString *joined=[files_strs componentsJoinedByString:@"\n"];
```
Comment (leaanthony) [inline v2/internal/frontend/desktop/darwin/WailsWebView.m:51]:
```suggestion
  NSString *joined=[files_strs componentsJoinedByString:@"\n"];
  
  // Release the array of file paths
  [files_strs release];
```

File: v2/internal/frontend/desktop/darwin/WailsWebView.m (hunk 2 above)
Comment (APshenkin) [reply v2/internal/frontend/desktop/darwin/WailsWebView.m:42]: Oh, yes you are right, sorry misunderstood you 😅

**Review by leaanthony — APPROVED** [review]

**Comment by leaanthony** [comment]
Thanks all involved in this monumental effort!

**Pull request merged**